### PR TITLE
fix #45825, BitArray methods assuming 1-indexing of AbstractArray

### LINKED ISSUE #45825: BitArray "constructor" incorrectly assumes 1-based indexing, leading to nondeterministic incorrect results due to out-of-bounds memory access
For example, in the following code, a `BitArray` that is constructed from an input array of values that are all `false` is found to contain some values that are `true`:

```jl
julia> using OffsetArrays

julia> all_false = falses(-1000:1000);

julia> any(==(true), all_false)
false

julia> all_false_bitarray = BitArray(all_false);

julia> any(==(true), all_false_bitarray)
true
```

This is due to the code [here](https://github.com/JuliaLang/julia/blob/2168230c210d620786311d91528a155582f78bc8/base/bitarray.jl#L510-L514), which applies `@inbounds` to a loop before proceeding to index the user-provided array with indices that start from one.

https://github.com/JuliaLang/julia/blob/2168230c210d620786311d91528a155582f78bc8/base/bitarray.jl#L510-L514

### PR DESCRIPTION
**Pull request by JeffBezanson**
fix #45825

### CODE REVIEW

File: base/bitarray.jl [hunk 1]
```diff
@@ -506,20 +506,23 @@ function BitArray{N}(A::AbstractArray{T,N}) where N where T
     Bc = B.chunks
     l = length(B)
     l == 0 && return B
-    ind = 1
+    Aind = eachindex(A)
```
Comment (N5N3) [inline base/bitarray.jl:509]: Local bench shows that this change makes this constructor almost 6x slower if `A` is linear-indexable. 
Should we move the copy part (L509~L528) to, e.g. `copyto!(dest::BitArray, src::AbstractArray)` and optimization for `src` with different `IndexStyle`?

**Comment by martinholters** [comment]
Might be too breaking before 2.0, but shouldn't `BitArray(::OffsetArray)` throw just like `Array{Bool}(::OffsetArray)` does (unless the start index happens to be 1)?

**Comment by JeffBezanson** [comment]
Oh I see, I was just wrong about what behavior is intended then. I was thinking of the argument as an iterator.

File: base/bitarray.jl (hunk 1 above)
Comment (JeffBezanson) [reply base/bitarray.jl:509]: What benchmark did you run?

Comment (N5N3) [reply base/bitarray.jl:509]:
```julia
julia> a = view(rand(Bool, 100, 100), :, :);

julia> @btime BitArray_master($a);
  1.340 μs (2 allocations: 1.38 KiB)

julia> @btime BitArray_pr($a);
  9.200 μs (2 allocations: 1.38 KiB)
```
`@code_llvm` shows that this PR failed to vectorize due to union split.

**Comment by JeffBezanson** [comment]
Ok this should be right now. Using `nextind` seems to give full performance.

File: base/bitarray.jl [hunk 2]
```diff
@@ -501,40 +501,40 @@ function Array{T,N}(B::BitArray{N}) where {T,N}
...
             Bc[i] = c
         end
         c = UInt64(0)
         for j = 0:_mod64(l-1)
-            c |= (UInt64(convert(Bool, A[ind])::Bool) << j)
-            ind += 1
+            c |= (UInt64(convert(Bool, A[Ai])::Bool) << j)
+            Ai = nextind(A, Ai)
         end
         Bc[end] = c
```
Comment (N5N3) [inline base/bitarray.jl:533]: L528 ~ L533 seems not correct
```julia
tail = _mod64(l)
c = Bc[nc] & (typemax(UInt64) << tail) # load the dest data and clean the bits which would be overwrote
for j = 0:tail-1
    c |= (UInt64(convert(Bool, A[Ai])::Bool) << j)
    Ai = nextind(A, Ai)
end
Bc[nc] = c # store back
```

**Pull request merged**